pkg/resource/restconfig_factory.go

```diff
@@ -1,31 +1,138 @@
 package resource
 
 import (
+	"context"
 	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	log "github.com/sirupsen/logrus"
 	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/restmapper"
 	"k8s.io/client-go/tools/clientcmd"
 	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"sync"
+	"time"
 )
 
+// HIVE-2399: In the cluster sync controller, getRESTConfigFactory is called once per
+// reconcile loop. Previously, a restConfigClientGetter was created for every reconcile,
+// used to apply (or rather, attempt to apply) various synced objects, and then thrown
+// away at the end of the reconcile pass.
+//
+// The restConfigClientGetter (or more accurately, the  cmdutil.Factory that wraps it) is
+// used by the cluster sync controller (via this helper) to apply the various objects to
+// the target cluster during the syncing process. During the application process (see apply.go),
+// the apply command is set up (aptly named setupApplyCommand()) using the factory returned
+// by the getRESTConfigFactory() below. Within this setup, the ToRESTMapper() function of the
+// restConfigClientGetter is invoked. While the mapper returned is a deferred variant,
+// eventually the inner NewDiscoveryRESTMapper() function is invoked, which is very expensive
+// and allocates a huge amount of memory over time if repeatedly called in a hot loop. A graphic
+// demonstrating these allocations is attached the HIVE-2399 card.
+//
+// The solution is twofold:
+// 1) store the deferred mapper on the restConfigClientGetter so that it can be reused
+// 2) cache the restConfigClientGetter for each rest config and reuse it for future reconciles
+// The type of this map is effectively map[string]*restConfigClientGetter
+var restConfigClientGetterCache sync.Map
+
+var cacheCleanMu sync.Mutex
+var lastCacheClean time.Time
+var cacheCleanInterval = time.Hour * 1
+var cacheEntryMaxAge = time.Hour * 8
+
+func cleanRestConfigClientGetterCache(logger log.FieldLogger) {
+	cacheCleanMu.Lock()
+
+	now := time.Now()
+	if now.Before(lastCacheClean.Add(cacheCleanInterval)) {
```

Comment:
So you're using this to keep this func an O(1) no-op the majority of the time. It's nice in that it's simple, but it's a tad hacky.

An alternative -- more complex but arguably more correcter -- is to make the cleanRestConfigClientGetterCache() func run fully async in its own queue, like the nameserver scraper. The queueing/polling behavior for that guy is defined [here](https://github.com/openshift/hive/blob/84c194aeb89c027a731fc1653081b1ca323b7a9d/pkg/controller/dnsendpoint/nameserverscraper.go#L211); and that Start function is implicitly run by the controller manager by having been Add()ed to it like [this](https://github.com/openshift/hive/blob/84c194aeb89c027a731fc1653081b1ca323b7a9d/pkg/controller/dnsendpoint/dnsendpoint_controller.go#L134).

Definitely not insisting on changing this, but wanted to put it in your head.